Revise SDL-0291 Allow navigation apps to access information about WiFi networks (#1073)
smartdevicelink/sdl_evolution: proposals/0291-allows-navigation-apps-to-access-information-about-Wi-Fi-networks.md

There is a complete solution to these problems.
It would help SDL greatly improve the user experience on `VideoStreaming` via BT+WiFi.

> To Listen to WiFi state change we register a receiver that receives WiFi state change to the transport layer,

Reviewer: ```suggestion
To listen to WiFi state change, we register a receiver that receives WiFi state change to the transport layer, and then modify the timing of request TCP connection when receiving WiFi state change.
```

> and then modify the timing of request TCP connection when receiving WiFi state change.

Reviewer: ```suggestion
```

                return;
            } else {
                Log.d(TAG, "onTransportDisconnected - " + disconnectedTransport.getType().name());
+               if (disconnectedTransport.getType() == TransportType.TCP && secondaryTransportParams != null) {
+                   if (activeTransports.containsValue(disconnectedTransport)
+                           && (transportManager != null && (transportManager.isWifiConnected() || transportManager.isWifiAPStateEnabled()))) {
+                       // If the established TCP connection is disconnected, the corresponding IP and port are invalid and should be removed from the list.

> +                       // Otherwise, isTransportForServiceAvailable is always true after disconnection

Reviewer: ```suggestion
+                       // Otherwise, isTransportForServiceAvailable is always true after disconnection.
```

> +                       // Do not remove when Wifi is connected or AP is enabled in HS side, because app need use it when connected again.

Reviewer: ```suggestion
+                       // Do not remove when WiFi is connected or access point is enabled on mobile device, because the app needs to use it when connected again.
```

         * @param packet a control frame packet
         */
        private void handleControlFrame(SdlPacket packet) {
            Integer frameTemp = packet.getFrameInfo();
            Byte frameInfo = frameTemp.byteValue();
...
            } else if (frameInfo == FrameDataControlFrameType.TransportEventUpdate.getValue()) {

                // Get TCP params
                String ipAddr = (String) packet.getTag(ControlFrameTags.RPC.TransportEventUpdate.TCP_IP_ADDRESS);
                Integer port = (Integer) packet.getTag(ControlFrameTags.RPC.TransportEventUpdate.TCP_PORT);

                if(secondaryTransportParams == null){
                    secondaryTransportParams = new HashMap<>();
                }

                if(ipAddr != null && port != null) {
                    String address = (port != null && port > 0) ? ipAddr + ":" + port : ipAddr;
                    secondaryTransportParams.put(TransportType.TCP, new TransportRecord(TransportType.TCP,address));

                    //A new secondary transport just became available. Notify the developer.
                    notifyDevTransportListener();
+               } else {

> +                   // Remove secondaryTransportParams when HU Wifi disconnected.

Reviewer: ```suggestion
+                   // Remove secondaryTransportParams when head unit WiFi is disconnected.
```

    }

+   public boolean isWifiConnected() {
+       return mIsWifiConnected;
+   }
+
+   public boolean isWifiAPStateEnabled() {
+       return mIsWifiAPStateEnabled;
+   }
...
    public interface TransportEventListener{
        /** Called to indicate and deliver a packet received from transport */
        void onPacketReceived(SdlPacket packet);

        /** Called to indicate that transport connection was established */
        void onTransportConnected(List<TransportRecord> transports);

        /** Called to indicate that transport was disconnected (by either side) */
        void onTransportDisconnected(String info, TransportRecord type, List<TransportRecord> connectedTransports);

        // Called when the transport manager experiences an unrecoverable failure
        void onError(String info);
        /**
         * Called when the transport manager has determined it needs to move towards a legacy style
         * transport connection. It will always be bluetooth.
         * @param info simple info string about the situation
         * @return if the listener is ok with entering legacy mode
         */
        boolean onLegacyModeEnabled(String info);

> +       /** Called to indicate that Wifi was connected/disconnected (by HS side) */

Reviewer: ```suggestion
+       /** Called to indicate that WiFi was connected/disconnected (by mobile device) */
```
